lib/plugins/aws/provider/awsProvider.js

```diff
@@ -204,10 +204,11 @@ class AwsProvider {
     const persistentRequest = (f) => new BbPromise((resolve, reject) => {
       const doCall = () => {
         f()
-          .then(resolve)
-          .catch((e) => {
-            if (e.statusCode === 429) {
-              that.serverless.cli.log("'Too many requests' received, sleeping 5 seconds");
+          .then(resolve, e => {
```

Comment:
We should not introduce the uncommon (and imo deprecated) 2-argument then signatures here, but stay with the commonly used `.then().catch()`.

Reply:
Hmm.. but the way it was, is not clean and actually confusing conceptually, that's why I patched it.

`p.catch(onError)` is equivalent to `p.then(null, onError)`. Therefore:

```javascript
p.then(onSuccess).catch(onFailure);
```
is equivalent to

```javascript
p.then(onSuccess).then(null, onError);
```

Writing chain that way, actually suggests that we use `onFailure` handler to also handle eventual error coming from `onSuccess` invocation, and this definitely not what we're after here.

While here crash of `resolve` is highly unlikely, I can imagine it may not be in other cases, therefore I would rather call it an anti-pattern especially if we do not plan to handle `onSuccess` crash with `onFailure` handler (we may accidentally silence error we do not intend to silence etc, or come with other wrong assumption).

I believe that in that case we want to purely _resolve on success and either recover or reject on failure_, and __not also:__ _either recover or reject on failure if resolve on success fails_ :)

Reply:
I think this is a general misconception introduced by the use of the `new Promise` anti-pattern in the function. Actually `.then(resolve)` can never fail as it just calls `Promise.resolve(result of f())` under the hood, so having a catch after the then will _only_ catch errors from the execution of `f()`. (You see that `f()` is actually executed before the `.then()` and errors will bypass the `then`). For this case the different signatures are sematically exactly equal.

What we should do is, get rid of the `new BbPromise` anti-pattern. Then the subsequent implementation of the done() semantics will clean up automatically and we'll end up with a clear and understandable implementation.

Reply:
> Actually .then(resolve) can never fail as it just calls Promise.resolve(successful result of f())

That's not how it works. Firstly `Promise.resolve` also can  also resolve with failure, e.g. run:

```javascript
Promise.resolve(Promise.reject(new Error("Some erorr"))).catch(err => {
   console.log(`I failed with: ${err.message}`);
});
```

And in `then` it happens more like:

```javascript
try {
  return Promise.resolve(fn());
} catch (e) {
  reutrn Promise.reject(e)
}
```

e.g. run following:

```javascript
new Promise(resolve => resolve()).then(() => {
  throw new Error("Some erorr");
}).catch(function (err) {
 console.log(`I failed with: ${err.message}`);
});
```

Reply:
Longer while ago created some educational material about promises, with step by step implementation: https://medikoo.com/asynchronous-javascript-interfaces/promises/?notes

While explanations stand in few place on`done` (which is not provided by native API). The described behavior and implementation of `then` and `catch` reflects  1:1 in how native implementation works

Reply:
Imo your example does not describe this specific implementation:

What we do here actually is:
```
return f()
.then(res => {
  resolve(res);    // This call can never fail and enter any catch path, because the `.then()` itself switches to catch
                    // and does not call the "then target function" in case f() rejects. This then target function is guaranteed to resolve 
                    // and not reject (as "resolve" is the native resolver function of the promise) and this function body is only executed if f() succeeds)
})
.catch(err => {
  // We will only end up here in case f() rejects
});
```
So the catch is only entered if f() rejects.

The notation
```
return f()
.then(res => {
  resolve(res);
}, err => {
  ...
});
```
does exactly the same, because the `then` calls the catch path if f() rejects.

Reply:
> So the catch is only entered if f() rejects.

But only with assumption that callback passed to  `then` can never fail :) and I agree that in this case this assumption is pretty safe, as what are the chances `resolve(res)`  will crash (?), not really possible.

Still it doesn't change the fact that it's logically flawed pattern. Firstly it doesn't resemble _if/else_ and (if logical purity is not what we're necessarily after) the more convincing argument should be that it's dangerous if repeated in not so "safe" cases.

e.g. see following example (it will crash with "API server error" message", even when API responds with success):

```javascript
apiRequest(url)
  .then(resultItems {
    resultItems.forEeach(...); // oops typo!
  }).catch(err =>  {
    throw Error("API server error, come back later!")
  });
```

So promoting `then(onSuccess).catch(onError)` as representation of _if/else_ handling is a quite bad idea.

---

Anyway main intention behind this PR is to improve recoverable AWS requests error, and not to fix imperfect (logically) promise flows. So if my arguments still remain not convincing enough for you, I can surely revert this promise flow patch

Reply:
> So if my arguments still remain not convincing enough for you, I can surely revert this promise flow patch

I agree with your argument that in general, the resemblance of a working if/then pattern as promise chain is important, as it can be copied without thinking from here 😄 .
So, we shall keep it. 

However, we also should add a comment that explains, why we decided to use the two-argument then here in favor of a then-catch construct, so that people looking at the code know the reason.

BTW: Good discussion 😄

Reply:
Good to hear :)

I added comment. I tried to keep it concise, but if you feel something more should be added, or reworded let me know.